### ISSUE TITLE
V0-C — prove a Vulkan Compute Unit through the InferSwarm execution seam

### ISSUE THREAD

**Issue by ezutfen** [Zutfen-LLC/inferswarm#143]
## Parent

#140 — V0 portable accelerator execution substrate evaluation

## Blocked by

#142 — V0-B evidence reduction and integration-seam selection

**DO NOT IMPLEMENT until #142 emits `V0B_PROCEED_TO_INTEGRATION_SPIKE` or the maintainer explicitly authorizes a narrower compatibility-tier spike.**

#142 must identify the exact runtime/execution seam this issue should use. This issue intentionally freezes architectural invariants and proof requirements now while leaving the backend-specific implementation shape evidence-driven.

## Objective

Prove the smallest doctrine-shaped InferSwarm integration in which a real Vulkan-capable accelerator is represented as an ordinary physical Compute Unit + Memory Resource with an observed Vulkan execution capability, and one strategy-defined execution unit is correctly materialized and executed there.

Core proof:

> A generic InferSwarm plan can select/realize/execute work on a Vulkan-capable Compute Unit because its capabilities, representation support, correctness qualification, and measured economics permit that plan—not because the generic planner knows the hardware is AMD, Vulkan, CUDA, ROCm, or any other vendor/backend class.

This is an **integration spike**, not production Vulkan support.

## Required predecessor handoff

Before implementation begins, #142 must provide:

- exact V0-A/V0-B evidence identities;
- exact physical Vulkan proving GPU/node;
- exact Vulkan driver/ICD/runtime identity to target initially;
- exact model/representation/workload class proven viable;
- correctness/stability findings;
- materialization/residency findings;
- prefill/decode economics;
- selected integration-seam class;
- reason that seam is preferable to the rejected alternatives;
- any backend-specific limitations that must become strategy/capability constraints rather than generic planner branches;
- a prospective correctness-comparison recommendation suitable for freezing before V0-C physical output is observed.

Record these as the V0-C **execution authority** before correctness-bearing physical execution.

If #142 cannot identify one bounded seam, stop. Do not invent one in implementation merely to make the issue progress.

## Architectural invariants

Preserve ADR 0008 / Fabric Doctrine semantics.

### Physical resource identity is backend-independent

The resource model must keep distinct concepts for:

- Node;
- Compute Unit;
- Memory Resource;
- discovered Link/topology relationships;
- execution backend capability/evidence.

Do not encode physical classes such as:

- `VulkanGPU`;
- `CudaWorker`;
- `AMDNode`;
- `ROCmTier`;
- `vulkan_primary` / `cuda_secondary`.

A physical Compute Unit may expose zero, one, or multiple viable execution backend capabilities.

### Strategy constrains; planner chooses

Model/backend-specific knowledge belongs at the strategy/backend boundary.

The generic planner may reason about opaque facts such as:

- execution implementation/capability identity;
- representation compatibility;
- required operations/features;
- memory capacity/headroom;
- measured correctness trust;
- measured startup/prefill/decode economics;
- policy/integrity eligibility.

It must not branch on model/backend/vendor nouns to choose the Vulkan device.

Forbidden generic logic includes equivalent forms of:

```text
if backend == vulkan
if vendor == amd
if gpu_is_cuda
if device_name contains "Radeon"
```

unless the code is strictly at a backend discovery/adapter boundary whose purpose is to report opaque generic capabilities upward.

### API remains unfrozen

Internal research structures may be introduced or refactored.

Do not declare a public backend plugin API, stable wire protocol, final capability schema, or production scheduler contract from this one spike.

## Scope

V0-C must prove **one bounded strategy-defined execution unit** on one real Vulkan Compute Unit.

The exact granularity is selected by #142's evidence. It may be, for example:

- a model block/stage;
- a bounded set of layers/state;
- another strategy-defined opaque execution unit;
- a whole small model only if #142 demonstrates that an opaque whole-model runtime is the only honest initial seam and the spike can still test resource/planner/backend separation.

Prefer the smallest unit that genuinely exercises:

1. physical resource/capability discovery;
2. representation/state materialization;
3. frozen plan realization;
4. Vulkan execution;
5. correctness comparison;
6. observed-state reconciliation;
7. generic planner/resource semantics.

Do not add distributed mixed-vendor execution merely for appearance.

## Cross-vendor resource graph requirement

The V0-C research resource snapshot should include, where safely available:

- the selected AMD/Vulkan Compute Unit from #141/#142;
- at least one known NVIDIA/CUDA Compute Unit from the existing fabric as a separate resource/capability observation.

This requirement is for **ontology/planning pressure**, not for forcing both devices into one execution plan.

Prove that both can coexist in one generic resource/evidence snapshot without one becoming an intrinsic primary/secondary/vendor tier.

If the strategy exposes legal candidates for both resources, retain them and let the generic eligibility/ranking path distinguish them from capabilities/evidence. If only one is technically legal for the frozen representation, the planner should exclude the other with a generic capability/representation explanation.

## Phase 0 — freeze V0-C methodology before physical output

Before correctness-bearing Vulkan execution, freeze:

- starting InferSwarm main SHA;
- exact #142 accepted/reviewed handoff identity;
- selected integration seam;
- exact proving node/GPU/BDF/device identity;
- Vulkan driver/ICD/runtime identity;
- exact execution-adapter/runtime source SHA and binary/container hashes;
- model/checkpoint/representation/state-unit identity;
- exact logical state covered by the execution unit;
- input fixtures;
- output observables;
- correctness comparator/methodology;
- any prospective numerical tolerance or semantic rule required for non-bit-exact backend comparison;
- generic resource/evidence snapshot schema used by the spike;
- legal candidate set;
- planner objective/policy sufficient for deterministic candidate selection;
- expected residency/materialization accounting;
- explicit non-claims.

Do not derive a tolerance from the same V0-C output it is used to judge. If #142 evidence is insufficient to freeze a defensible numerical comparator, V0-C may first run a clearly separate calibration/methodology subgate or stop for a new qualification-method issue.

## Phase 1 — backend capability discovery

Implement the narrowest discovery/adapter path necessary to expose the selected Vulkan execution implementation as evidence associated with the physical Compute Unit.

At minimum the research representation must distinguish:

### Physical facts

- node identity;
- physical Compute Unit identity;
- Memory Resource identity/capacity;
- topology/link facts.

### Execution-capability facts

- backend/implementation identity;
- driver/runtime identity;
- supported representation(s);
- required/available execution features relevant to the frozen unit;
- measured correctness trust/evidence reference;
- measured economics reference;
- evidence freshness.

Do not infer capability from GPU marketing name where it can be observed directly.

Negative controls must prove a Vulkan capability cannot be attached to the wrong physical device/evidence identity without rejection.

## Phase 2 — strategy candidate exposure

Extend or use the current Model Execution Strategy boundary so the frozen logical unit exposes legal implementation alternatives without teaching the generic planner Vulkan semantics.

For each candidate retain at least:

- opaque candidate/execution-implementation identity;
- logical state/execution units covered;
- required representation;
- required generic capabilities/features;
- memory requirement/headroom;
- correctness/evidence requirements;
- relevant measured economics;
- legal state/materialization boundary.

If both Vulkan and vendor-native candidates are legal, expose both.

If the V0-C representation is legal only on Vulkan or only on a subset of resources, make the generic required-capability reason explicit.

No manual `use_vulkan=true` choice may substitute for the planner proof in the canonical run. A diagnostic override is allowed only for pre-canonical debugging and must be clearly marked non-authoritative.

## Phase 3 — generic planning

Use the existing generic planning seam where practical.

At minimum prove:

1. the strategy exposes the frozen legal candidate set;
2. technical feasibility runs before optimization;
3. integrity/correctness/evidence eligibility is explicit;
4. memory capacity/headroom is enforced;
5. measured evidence applicability is explicit;
6. ranking/tie-break behavior is deterministic under the frozen objective;
7. selection/exclusion explanations contain generic capability/evidence/resource reasons;
8. no generic planner code branches on Vulkan/CUDA/AMD/NVIDIA/model-family nouns.

If V0-C discovers that the current planner representation cannot express a genuinely generic backend execution alternative, classify the necessary change as either:

- `GENERIC_BUT_REQUIRED_EXTENSION`; or
- `BACKEND_LEAKAGE_REFACTOR_REQUIRED`.

Do not hide the problem with a Vulkan-specific planner exception.

## Phase 4 — selective materialization / residency

Realize only the state required by the frozen execution unit plus explicitly declared shared/runtime state.

Retain:

- source/backing identity;
- state selected before materialization where applicable;
- staging bytes;
- final device-local resident bytes;
- required persistent host bytes;
- intentionally retained host cache bytes;
- released staging bytes;
- unexplained persistent host mirror bytes;
- representation/conversion buffers;
- observed GPU memory before/after realization.

Preserve the existing doctrine requirement:

> accelerator residency must not inherently require an unexplained equivalent persistent host-RAM mirror.

A backend/runtime may deliberately require persistent host state, but it must be explicit and economically accounted rather than hidden.

For a passing V0-C retained plan:

`unexplained_persistent_host_mirror_bytes == 0`

unless #142 explicitly established that the selected seam cannot satisfy this and the maintainer authorized a narrower diagnostic spike. Such a diagnostic cannot claim full V0-C pass.

## Phase 5 — real Vulkan execution

Execute the frozen strategy unit on the exact selected physical Vulkan device.

Mechanically prove:

- exact plan/candidate identity;
- exact physical device/backend identity;
- exact logical input/state identity;
- intended state is accelerator resident;
- the Vulkan backend performed the correctness-bearing operation;
- no silent CPU/CUDA/HIP/other-GPU fallback;
- output is attributed to the correct plan/operation/unit;
- no unplanned model-state movement occurs during steady-state execution;
- no unplanned source fetch/rematerialization occurs;
- no NaN/Inf;
- explicit failure rather than silent substitution when the backend cannot execute.

Use the correctness methodology frozen in Phase 0.

Performance victory is not required for V0-C architectural success, but record enough wall/throughput/memory data to verify #142's economics remain directionally applicable under the actual integration seam.

## Phase 6 — control and negative proofs

At minimum include controls for:

1. wrong physical-device binding;
2. stale/wrong V0-A/V0-B evidence identity;
3. unsupported representation;
4. insufficient device memory/headroom;
5. required backend capability missing;
6. backend initialization failure;
7. model/state identity mismatch;
8. corrupted or mismatched materialization;
9. attempted silent CPU/native-backend fallback;
10. planner candidate with missing/insufficient correctness trust;
11. execution result attributed to the wrong plan/unit/device;
12. generic planner source audit proving no Vulkan/AMD/CUDA/NVIDIA selection branch was introduced.

All must fail closed or produce an explicitly lower-trust/ineligible candidate as appropriate.

## Phase 7 — mixed-resource ontology proof

With the research resource snapshot containing at least the selected Vulkan Compute Unit and one native CUDA Compute Unit where safely available, prove:

- both resources use the same generic physical ontology;
- backend capabilities/evidence are associated facts, not subclasses/tiers;
- a generic explanation can state why each resource was selected, lower-ranked, infeasible, or unused;
- adding the Vulkan resource does not require changing the semantics of an existing CUDA Compute Unit;
- removing the Vulkan capability/evidence causes generic replanning/exclusion rather than a vendor-special-case crash.

This is not a requirement to execute one request across both vendors yet.

## Assumption audit

Produce an explicit audit of the integration against the current architecture using:

- `GENERIC_AND_REUSED_UNCHANGED`;
- `GENERIC_BUT_REQUIRED_EXTENSION`;
- `BACKEND_LEAKAGE_REFACTORED`;
- `FOUNDATIONAL_INVARIANT_FALSIFIED`.

Audit at least:

- resource graph;
- Memory Resource representation;
- execution capability/evidence representation;
- strategy boundary;
- generic candidate representation;
- planner eligibility/ranking;
- frozen plan;
- materialization/residency;
- observed-state reconciliation;
- execution/result attribution.

If a foundational doctrine invariant is actually falsified, stop and surface it for an ADR/Fabric Doctrine decision rather than declaring V0-C complete.

## Evidence retention

Create an additive area such as:

`docs/investigations/vulkan-v0-c/`

Retain at minimum:

- #142 handoff/authorization identity;
- methodology frozen before physical execution;
- resource/evidence snapshot;
- backend capability records;
- strategy legal candidate set;
- planner decision/explanations;
- frozen plan;
- materialization/residency accounting;
- exact physical/backend execution identity;
- correctness raw observations/reduction;
- fallback/source-fetch/state-movement counters;
- negative-control results;
- mixed-resource ontology proof;
- assumption audit;
- performance sanity measurements;
- explicit non-claims;
- manifests/hashes.

### `V0C_VULKAN_EXECUTION_SEAM_PASS`

Use only when retained evidence proves:

- #142 authorized the exact seam;
- one real Vulkan Compute Unit is represented through generic resource semantics;
- backend capability/evidence remains separate from physical resource identity;
- strategy exposes legal implementation candidate(s) without leaking Vulkan policy into the generic planner;
- generic planning selects/excludes deterministically from capability/evidence/resource facts;
- required state is materialized/resident with truthful memory accounting and zero unexplained persistent host mirror bytes;
- the correctness-bearing operation executes on the intended Vulkan device without fallback;
- the prospectively frozen correctness contract passes;
- observed-state/result attribution is clean;
- negative controls fail closed;
- CUDA and Vulkan resources can coexist in the same generic resource/evidence snapshot;
- no foundational architecture invariant remains falsified and undocumented;
- no public API is declared stable.

### `V0C_SEAM_REFACTOR_REQUIRED`

Use when Vulkan execution itself is viable but the attempted integration exposes a generic/backend-boundary defect that must be refactored before a valid pass can be claimed.

### `V0C_VULKAN_INTEGRATION_FAIL`

Use when the #142-selected path cannot execute/materialize correctly or cannot preserve required InferSwarm invariants.

### `V0C_EVIDENCE_BLOCKED`

Use when prerequisites/correctness methodology/physical execution cannot support a trustworthy result.

## Explicit non-goals

V0-C does not require or claim:

- production Vulkan support;
- Vulkan as default/preferred backend;
- production mixed AMD/NVIDIA serving;
- simultaneous cross-vendor execution in one request;
- multi-node Vulkan serving;
- final planner/backend/plugin APIs;
- automatic driver installation/management;
- every Vulkan model/quantization;
- performance parity with CUDA/ROCm;
- fault-tolerant backend failover;
- user-facing configuration/UI.

Those require separate evidence-gated follow-ups after #140 is resolved.

## Acceptance criteria

- [ ] #142 explicitly authorizes the selected V0-C seam before implementation/physical execution.
- [ ] Prospective correctness methodology frozen before V0-C output is observed.
- [ ] Real Vulkan physical Compute Unit + Memory Resource represented independently of backend identity.
- [ ] Vulkan capability/evidence mechanically bound to the correct physical device.
- [ ] Strategy-defined legal candidate exists without generic planner vendor/backend branching.
- [ ] Generic planner selects/excludes using resource/capability/evidence facts.
- [ ] Frozen plan realizes only required state with truthful staging/residency accounting.
- [ ] `unexplained_persistent_host_mirror_bytes == 0` for a PASS.
- [ ] Real correctness-bearing execution occurs on Vulkan with no silent fallback.
- [ ] Frozen comparator passes for a PASS.
- [ ] Negative controls fail closed.
- [ ] Mixed Vulkan + CUDA resources coexist in one generic resource/evidence snapshot without intrinsic tiers.
- [ ] Assumption audit completed.
- [ ] Existing non-Vulkan regressions remain green for every generic seam changed.
- [ ] No public API or preferred-backend ADR is frozen from this spike.
- [ ] Accepted Issue #117 evidence/authority remains untouched.

## Completion report

Include:

1. starting main SHA;
2. exact #142 handoff/terminal consumed;
3. selected seam and why #142 chose it;
4. exact proving node/GPU/BDF/backend/driver identity;
5. files/modules changed;
6. generic resource/capability representation used;
7. strategy legal candidates;
8. planner decision and no-backend-branch proof;
9. frozen plan identity;
10. state/materialization/residency accounting;
11. correctness methodology and result;
12. fallback/source-fetch/state-movement counters;
13. negative-control results;
14. mixed-resource ontology proof;
15. assumption audit;
16. focused/full/exact-head CI results;
17. retained evidence paths/manifests;
18. exact V0-C terminal;
19. recommendation to parent #140;
20. confirmation PR remains **OPEN / UNMERGED** for maintainer review.

Do not merge the implementation PR or promote Vulkan policy without explicit maintainer action.

**Comment by ezutfen** [comment]
## Maintainer authorization — V0-C agent handoff

#142 is ACCEPTED and complete. PR #146 merged to `main` as `d9bcd539e2191056a9340cb69f378dfa018bb8de` from accepted final V0-B head `c912891c669412d5206697bb537eb77f27f55824`; exact-head CI run `34702050860` succeeded.

This satisfies this issue's implementation gate. V0-C is now authorized to proceed, subject to this issue's mandatory Phase-0 freeze before any correctness-bearing physical execution.

Consume the accepted V0-B handoff exactly as follows:

- terminal: `V0B_PROCEED_TO_INTEGRATION_SPIKE`;
- selected seam: `S2-backend-adapter-participant`;
- generic planner remains backend-neutral above the execution boundary;
- no public API freeze, no planner backend nouns, no preferred/default-backend ADR;
- S2 must coexist with CUDA and HIP capabilities under the generic resource graph;
- AMD-A/Vulkan backend-local repeatability is established by 3/3 identical corrected visible generations;
- AMD-B/Vulkan, NV-A/Vulkan, and NV-A/CUDA generated-output repeatability remain `insufficiently_observed` (single corrected observation each); do not upgrade them;
- NV-A CUDA/Vulkan remains a `generated-output difference`; no semantic equivalence or logits-level numerical equivalence is accepted;
- ADR-0010 layer-2 numerical equivalence and layer-3 strategy-declared semantic correctness remain `NOT ESTABLISHED`; V0-C must freeze its correctness methodology prospectively before canonical physical output;
- the supplemental CPU arm is retrospective/descriptive context only. Retained bytes factually support `layers-executed-on-host-CPU`, but correction 3 was written after collection, so the CPU baseline is NOT prospectively frozen decision-grade authority and must not be used as a V0-C eligibility gate or acceptance threshold;
- AMD throughput causal decomposition remains open (compute vs memory vs link); do not encode the Polaris representation/capability observations as a causal performance claim;
- accepted V0-A authority remains PR #145 merge `273b9e8e32c779f063903cd75a0c6772d0d1e451`, V0-A evidence head `2b2d546dc77350f14016b31a105696b257c063f2`.

Start from current `main` at/after merge `d9bcd539e2191056a9340cb69f378dfa018bb8de`. Reconcile main before work. Freeze the exact #142 handoff identity, selected S2 seam, proving resource/backend identities, representation/state unit, legal candidates, planner policy, residency accounting, and prospective correctness comparator before correctness-bearing Vulkan execution.

Issue #143 is now agent-ready. Leave its implementation PR OPEN/UNMERGED for maintainer review.
